### PR TITLE
Exercise 1 - Word Counter - Develop a routine for counting words in a file

### PR DESCRIPTION
**Pull request by robertapplin**
This PR adds a Word Counter program. 
The Word Counter program count's the usage of words, which are over 4 characters long, from a provided file and output's these words in order of most used to least used.

### CODE REVIEW

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using WordCount = std::vector<std::pair<std::string, std::size_t>>;
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: Good. Using a type alias here makes a lot of sense. 👍

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * @return An iterator pointing to the next letter contained
+ *         in the supplied set.
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: Good that you described the `@return` of the function but what about the `@params`s? And a summary of what the function does.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+ */
+std::string::const_iterator
+findNextCharacter(std::string::const_iterator fromIterator,
+                  std::string::const_iterator toIterator,
+                  const std::string &characters) {
+  for (auto it = fromIterator; it < toIterator; ++it) {
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp:30]: Perhaps use `const auto` for clearer intent?

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 2 above)
Comment (samueljackson92) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: Also update other places where this is true.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 3 above)
Comment (samueljackson92) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp:30]: Also update other places where this is true.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+ */
+bool compareCounts(const std::pair<std::string, std::size_t> &count1,
+                   const std::pair<std::string, std::size_t> &count2) {
+  return count1.second > count2.second;
+}
+
+/**
+ * @return True if the string is shorter than 5 characters
+ */
+bool isShorterThanFive(const std::string &str) { return str.length() < 5; }
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: This is pretty specific. Perhaps just move this to be a lambda function:

```cpp
auto it = std::remove_if(strings.begin(), strings.end(), [](const std::string& str) {
    return str.length() < 5;
});
```

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  // Remove invalid indices
+  strings.erase(it, strings.end());
+}
+
+/**
+ * @return A string transformed to lower case
+ */
+std::string toLower(std::string str) {
+  std::transform(str.begin(), str.end(), str.begin(),
+                 static_cast<int (*)(int)>(std::tolower));
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: Prefer a lambda over this weird cast to a function pointer:
```cpp
 std::transform(str.begin(), str.end(), str.begin(), [](const char c){ std::tolower(c); });
```

It should equate to the same thing in any decent compiler.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * Splits the string into each word, removes short words and changes
+ * each word to lower case before counting the number of occurances.
+ */
+void countWords(const std::string &str, const std::string &delimiters,
+                std::unordered_map<std::string, std::size_t> &count) {
+  auto words = splitBy(str.begin(), str.end(), delimiters);
+  removeShorterThanFive(words);
+  std::transform(words.begin(), words.end(), words.begin(), toLower);
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: You wrote a `toLower` function, but you haven't used it here.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+/**
+ * @return A vector pair containing each word paired to their usage count.
+ */
+WordCount countWords(std::istream &start, const std::string &delimiters) {
+  std::unordered_map<std::string, std::size_t> count;
+  for (std::string line; std::getline(start, line);)
+    countWords(line, delimiters, count);
+  return std::vector<std::pair<std::string, std::size_t>>(count.begin(),
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: `return WordCount(...)` instead?

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * @return A vector pair containing each word paired to their usage count.
+ */
+WordCount countWordsInFile(const std::string &filename,
+                           const std::string &delimiters) {
+  std::ifstream filestream(filename);
+  if (!filestream)
+    throw std::runtime_error("Error occured when attempting to load file: " +
+                             filename);
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: You did not remember to close the file stream before throwing an exception. While this is a trivial toy example I would suggest using [RAII](https://en.cppreference.com/w/cpp/language/raii) instead.

Maybe write yourself a tiny file reading class to do this.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 1 above)
Comment (samueljackson92) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: I would consider perhaps even type aliasing the `std::pair` as well as this is used in several places.

**Review by samueljackson92 — COMMENTED** [review]
Generally very good. I have a few minor comments listed below. 

Nice to see a good use of standard algorithms throughout the code. It is also generally clean and well formatted. Good work.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 7 above)
Comment (robertapplin) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: Ahh yes, that would have been much better!

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 6 above)
Comment (robertapplin) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: I think I have used it there

Comment (samueljackson92) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: Yes but I think you could have done:
```cpp
words = toLower(words)
```

?

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 5 above)
Comment (robertapplin) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: That looks much better!

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 6 above)
Comment (robertapplin) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp]: Yes, you are right!

File: exercises/Robert_Applin/ex01_basics/src/main.cpp (hunk 3 above)
Comment (robertapplin) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp:30]: Good point, but it will be changing, so it can't be const. Is that right?

Comment (samueljackson92) [reply exercises/Robert_Applin/ex01_basics/src/main.cpp:30]: Ah apologies. I think you're correct. The iterator needs to be mutable here. Forget this comment.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 9]
```diff
@@ -13,11 +13,15 @@
 #include <unordered_set>
 #include <vector>
 
-using WordCount = std::vector<std::pair<std::string, std::size_t>>;
+using WordCountPair = std::pair<std::string, std::size_t>;
+using WordCount = std::vector<WordCountPair>;
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: 👍

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 10]
```diff
@@ -13,11 +13,15 @@
...
 
 /**
+ * function finds the end point of a word.
+ * @params Iterators representing the start and end points of an
+ *         iteration. The characters used to split up each word.
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: 👍

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 11]
```diff
@@ -63,39 +73,42 @@ std::vector<std::string> splitBy(std::string::const_iterator fromIterator,
...
- * Removes any words shorter than five from the vector
+ * function removes any words shorter than five from the vector
+ * @params A vector of words
+ * @return void
  */
 void removeShorterThanFive(std::vector<std::string> &strings) {
-  auto it = std::remove_if(strings.begin(), strings.end(), isShorterThanFive);
+  auto it =
+      std::remove_if(strings.begin(), strings.end(),
+                     [](const std::string str) { return str.length() < 5; });
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: 👍

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 12]
```diff
@@ -63,39 +73,42 @@ std::vector<std::string> splitBy(std::string::const_iterator fromIterator,
...
 /**
+ * function transforms a string to lower case
+ * @params A string representing a word
  * @return A string transformed to lower case
  */
 std::string toLower(std::string str) {
   std::transform(str.begin(), str.end(), str.begin(),
-                 static_cast<int (*)(int)>(std::tolower));
+                 [](const char str2) { return std::tolower(str2); });
   return str;
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: 👍

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 13]
```diff
@@ -117,44 +131,57 @@ void countWords(const std::vector<std::string> &substrings,
...
+ *         around and a map containing each word and their count
+ * @return void
  */
 void countWords(const std::string &str, const std::string &delimiters,
                 std::unordered_map<std::string, std::size_t> &count) {
   auto words = splitBy(str.begin(), str.end(), delimiters);
   removeShorterThanFive(words);
-  std::transform(words.begin(), words.end(), words.begin(), toLower);
+  for (auto &word : words) {
+    word = toLower(word);
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: I take it back. I thought here we were operating on a string, not a vector. You were correct to prefer the `std::transform`. My apologies it was late on Friday.

File: exercises/Robert_Applin/ex01_basics/src/main.cpp [hunk 14]
```diff
@@ -117,44 +131,57 @@ void countWords(const std::vector<std::string> &substrings,
...
+ * function attempts to access the provided file
+ * @params The file name, and the characters each word will be split around.
  * @return A vector pair containing each word paired to their usage count.
  */
 WordCount countWordsInFile(const std::string &filename,
                            const std::string &delimiters) {
   std::ifstream filestream(filename);
-  if (!filestream)
+  if (!filestream) {
+    filestream.close();
```
Comment (samueljackson92) [inline exercises/Robert_Applin/ex01_basics/src/main.cpp]: 👍

**Review by samueljackson92 — APPROVED** [review]

**Pull request closed without merging**